### PR TITLE
fix(no-unnecessary-type-assertion): enable skipped upstream cases

### PR DESCRIPTION
**Pull request by camc314**
This follows up the upstream test sync by enabling the cases that were temporarily skipped.

The implementation now mirrors the upstream rule more closely for the cases those tests cover: same-type checks account for structural and generic edge cases, contextual fallback handles unnecessary assertions accepted by the receiver, double assertions can be removed when the original expression is already sufficient, and non-null assertions follow the newer constrained-vs-actual type behavior.

The goal is to keep the synced tests active without relaxing coverage, while preserving the existing false-negative choices around overloaded calls, generic inference, phantom type arguments, logical assignments, satisfies expressions, and implicit-void IIFEs.

fixes https://github.com/oxc-project/tsgolint/pull/1002

### CODE REVIEW

File: internal/rules/no_unnecessary_type_assertion/no_unnecessary_type_assertion.go [hunk 1]
```diff
@@ -167,58 +167,243 @@ var NoUnnecessaryTypeAssertionRule = rule.Rule{
...
+			}
+			seen[t] = struct{}{}
+			if predicate(t) {
 				return true
 			}
+			for _, part := range append(utils.UnionTypeParts(t), utils.IntersectionTypeParts(t)...) {
+				if part != t && typeContains(part, predicate, seen) {
+					return true
+				}
+			}
```
[the remark this answers is not in the document]
Comment (camc314) [reply internal/rules/no_unnecessary_type_assertion/no_unnecessary_type_assertion.go]: Good catch. I split the union and intersection constituent traversal into separate loops so we no longer append to a slice that may share backing storage with `t.Types()`. Verified with `go test ./internal/rules/no_unnecessary_type_assertion/...`.

**Comment by camc314** [comment]
### Merge activity

* **Jun 3, 10:03 AM UTC**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.com/merges?org=oxc-project&repo=tsgolint) once it meets the requirements.
* **Jun 3, 10:09 AM UTC**: `camc314` added this pull request to the [Graphite merge queue](https://app.graphite.com/merges?org=oxc-project&repo=tsgolint).
* **Jun 3, 10:16 AM UTC**: Merged by the [Graphite merge queue](https://app.graphite.com/merges?org=oxc-project&repo=tsgolint).
